core: Provide read access to commit/message log and odb

kim opened the pull request:
# Description of Changes

This is a first approximation to provide data access for Kafka-style replication. It punts on notifications of segment rotation (which could also be solved via filesystem events).

ObjectDB access is fairly crude, and will likely require it's own replication subsystem.


~~Note that this is against `cloud-next`.~~

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

kim committed:
core: Provide read access to commit/message log and odb

This is a first approximation to provide data access for Kafka-style
replication. It punts on notifications of segment rotation (which
could also be solved via filesystem events).

ObjectDB access is fairly crude, and will likely require it's own
replication subsystem.

kim commented:
This will probably do, so rebased into `master`

cloutiertyler commented on crates/core/src/db/message_log.rs:
```diff
@@ -203,6 +204,55 @@ impl MessageLog {
+    /// The iterator represents a _snapshot_ of the log at the time this method
+    /// is called. That is, segments created after the method returns will not
+    /// appear in the iteration. The last segment yielded by the iterator may be
+    /// incomplete (i.e. still be appended to).
```
These doc comments are A+ 🙏

cloutiertyler commented on crates/core/src/db/commit_log.rs:
```diff
@@ -121,3 +127,150 @@ impl CommitLog {
+
+/// A read-only view of a [`CommitLog`].
+pub struct CommitLogView {
+    mlog: Option<Arc<Mutex<MessageLog>>>,
```
I might suggest implementing the `CommitLog` by giving it a `CommitLogView` internally, since it does already share the same fields. Hmm that doesn't seem like the right solution either, but I wonder if we can have a clearer abstraction here.

It seems like we are almost trying to create the concept of a "File" (i.e. `CommitLogView`) and a "BufferedFile" (i.e. `CommitLog`) where the buffered thing is unwritten commit which is only periodically flushed. I wonder if something like that might fit better down the road.

kim replied:
The purpose of `CommitLogView` is for `RelationalDB` to hand out a view over its _internal state_ in order to serve up log segments. It must hide `append`, but changing the visibility of `append` isn't helpful.

`CommitLogView` could be thought of as a newtype around `CommitLog`, but internally that'd be copying more data than strictly required. Perhaps if there was an `Arc` around the whole `CommitLog`, but I'm always wary about contention.

cloutiertyler commented on crates/core/src/db/message_log.rs:
```diff
@@ -217,6 +267,124 @@ impl MessageLog {
+/// The underlying file is opened lazily when calling [`SegmentView::try_into_iter`]
+/// or [`SegmentView::try_into_file`].
+#[derive(Clone, Debug)]
+pub struct SegmentView {
```
Why is this required instead of `MessageLogIter`? Also presumably we should implement the internals of `MessageLogIter` with `SegementView` because we're duplicating some code that really needs to stay in sync.

kim replied:
I was going to ask about `MessageLogIter`, and perhaps submit a separate patch: the issue is, it has very different semantics. For one, it panics on I/O errors, which isn't great when it's not a matter of database consistency. More importantly, though, `segments_for_offset` returns the **first** segment if `offset` does not fall into the **last** segment -- the significance of which I don't understand.

So imho this would require to write a number of tests at least, which I wouldn't want to conflate with this patch (which does not change any database internals).

kim replied:
> Why is this required instead of MessageLogIter?

It is also not helpful to borrow `MessageLog` when that value is behind a mutex.

cloutiertyler reviewed (commented):
More or less seems okay, although I have some important questions inline.

kim committed:
Merge remote-tracking branch 'origin/master' into kim/log-access

kim committed:
fixup! Merge remote-tracking branch 'origin/master' into kim/log-access

kim committed:
Add some tests

Also make max segment size configurable, so tests don't have to write
>= 1GiB worth of data.

kim committed:
Abide to the Rust naming conventions

kim committed:
Add test for commit iter

kim committed:
Remove `MessageLogIter`, use `commit_log::Iter` instead (#272)

cloutiertyler reviewed (approved):
LGTM ultimately, although I'd like to think about a different philosophy on this
